Split chat lines into pools and allow the player to filter them

dragunoff opened the pull request:
This is #17737 stripped down to the core idea:
- Split chat lines into pools (e.g. System, Mission, Chat, Notifications, UI Feedback)
- Allow the player to toggle the display of UI Feedback chat pool

This is the only UI exposed to the player - a checkbox in "Display settings":
![Screenshot_20210208_195234](https://user-images.githubusercontent.com/1355810/111044029-c73e5a80-844e-11eb-9c90-cfed0631e822.png)
![Screenshot_20210208_195512](https://user-images.githubusercontent.com/1355810/111044030-c86f8780-844e-11eb-9ad0-ca7e34a1c8a1.png)

_Display settings_

Repeated messages are merged into one and a counter is displayed:
![image](https://user-images.githubusercontent.com/1355810/101391120-8e7d6b00-38cc-11eb-963a-9b3c5a3170fa.png)
_In game chat_

Closes #16148
Provides the basis for solving #5775, #6400, #6354, #3278 and #12421. These will come in follow-ups.

dragunoff committed:
Split chat lines into pools

- Add a common class for passing around chat lines
- Add wrapper methods for adding chat lines
- Combine repeated chat lines in the display widget

dragunoff committed:
Allow the player to toggle the display of UI Feedback chat pool

anvilvapre commented:
at the end of a long game i tried to read back at what play time another player left the game. when who said what. i noticed the chat lines did not contain a timestamp. perhaps a separate feature. not important as you may be able to see in replay.

dragunoff commented:
Needs a rebase but I'll wait for #18948

dragunoff commented:
Rebased and updated to work with the new settings logic split.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -891,14 +891,30 @@ public static void AddSystemLine(string text)
+			AddChatLine(name, nameColor, text, chatMessageColor, ChatPool.Chat);
+		}
+
+		public static void AddChatLine(string name, Color nameColor, string text, Color textColor,  ChatPool pool)
```
Whitespace error before `ChatPool`

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -891,14 +891,30 @@ public static void AddSystemLine(string text)
+			AddChatLine(name, nameColor, text, chatMessageColor, ChatPool.Chat);
+		}
+
+		public static void AddChatLine(string name, Color nameColor, string text, Color textColor,  ChatPool pool)
```
Also IMO we now have way too many (7!) different methods for adding chat lines. How disruptive would it be to cut this down to require all callers to specify their pool (and IMO move that to the first argument).

Maybe something like this?
```csharp
public static void AddChatLine(ChatPool pool, string text, Color? textColor = null);
public static void AddChatLine(ChatPool pool, string prefix, string text, Color? prefixColor = null, Color? textColor = null);
```
Don't force things into these specific signatures if these are not the most natural for our use cases.

Shouldn't we also be doing the `ChatPoolFilters` check here? Code that calls this directly with `ChatPool.Feedback` will ignore filtering.

dragunoff replied:
Yes, I also think there are too many methods. I talked about this in the previous PR (see https://github.com/OpenRA/OpenRA/pull/17737#discussion_r460856381). I like the idea of requiring callers to specify a pool :+1: I will try to improve this.

dragunoff replied:
Ugh, this needs more work but I definitely want to base it on #19360

dragunoff replied:
I went over this a few times and after all I like it better with separate methods than requiring a pool to be specified each time. Having several methods is less of a problem now that chat handling is no longer cluttering `Game.cs`. I changed the method signature a bit after your suggestion settling on this:
```csharp
public static void AddChatLine(string prefix, string text, Color? prefixColor = null, Color? textColor = null, TextNotificationPool? pool = null);
```

pchote commented on OpenRA.Game/Network/ChatLine.cs:
```diff
@@ -0,0 +1,55 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
```
2021 now

pchote commented on OpenRA.Game/Network/ChatLine.cs:
```diff
@@ -0,0 +1,55 @@
+			Pool = chatPool;
+		}
+
+		public bool IsRepeatable()
```
IMO its not clear what "repeatable" actually means. How about making this a `RemoveOnDuplicate` (or similar, if my interepretation is wrong) property instead?

dragunoff replied:
Well, it's not just removing on duplicate because there's a counter being incremented and appended to the message (e.g. "Unit lost (5)"). But I agree that `IsRepeatable` is not be the best name.

dragunoff replied:
Renamed to `IncrementOnDuplicate`.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -231,38 +234,55 @@ public void CloseChat()
 		}
 
-		void AddChatLine(string @from, Color nameColor, string text, Color textColor, bool suppressSound)
+		void AddChatLine(ChatLine chatLine, bool isRepeated, bool suppressSound)
```
`isRepeated` doesn't sit right with me here either. In practice what this is actually doing is removing the last widget in the scroll panel before adding the new, so wouldn't it be better to just do this explicitly in `AddChatLineWrapper` before calling this?

dragunoff replied:
Yeah, `AddChatLineWrapper` is a better place to handle this :+1:

dragunoff replied:
I moved the checks for repeated lines to `AddChatLineWrapper`.

pchote commented on OpenRA.Game/Settings.cs:
```diff
@@ -33,6 +33,13 @@ public enum MPGameFilters
 	}
 
+	[Flags]
+	public enum ChatPoolFilters
```
These needs to be defined in the first commit to avoid a compile error.

pchote reviewed (commented):
A few code comments below. Ingame, the message deduplication is really nice. I wonder whether we need a tooltip on the checkbox to describe what "UI Feedback" actually covers?

dragunoff commented:
> A few code comments below. Ingame, the message deduplication is really nice. I wonder whether we need a tooltip on the checkbox to describe what "UI Feedback" actually covers?

Yeah, "UI Feedback" and the pool associated with it is a bit problematic. I had difficulty coming up with a name for that pool. Currently it only includes the "Selected across screen/map" messages and they probably don't belong in the chat at all. There was an idea to move them to a dedicated widget for transient messages (it was discussed in #16355). And perhaps that's a good way to handle them.

dragunoff commented:
Another thing that may be a problem with the "repeatable" lines that I introduce in this PR is the number appended to the end. Because "Selected across map (3)"  is a somewhat ambiguous - does that mean that 3 units were selected or that the message was repeated 3 times? :thinking:

dragunoff commented:
Update: I'm trying to get back to this but I should wait for #19360 (or rebase on top of it)  because it moves the chat handling functions out of `Game.cs`.

teinarss commented:
Needs a rebase now.

abcdefg30 commented on OpenRA.Game/Network/ChatLine.cs:
```diff
@@ -0,0 +1,55 @@
+			Pool = chatPool;
+		}
+
+		public bool IncrementOnDuplicate()
```
```suggestion
		public bool CanIncrementOnDuplicate()
```
`IncrementOnDuplicate` sounds like it will do the increment.

dragunoff commented:
Update:
- rebased after the merge of #19436 
- renamed `ChatLine` to `TextNotification` to go better with the new `TextNotificationManager` class
- applied other changes and fixes from the discussions above

teinarss commented on OpenRA.Game/Network/TextNotification.cs:
```diff
@@ -0,0 +1,56 @@
+using System;
+using OpenRA.Primitives;
+
+namespace OpenRA.Network
```
Can we move this out from the Network folder and namespace?

dragunoff replied:
Moved to the main namespace and the main folder.

teinarss commented on OpenRA.Game/TextNotificationsManager.cs:
```diff
@@ -28,24 +29,44 @@ static TextNotificationsManager()
+		{
+			var filters = Game.Settings.Game.TextNotificationPoolFilters;
+
+			return pool == TextNotificationPool.Chat ||
+				pool == TextNotificationPool.System ||
+				pool == TextNotificationPool.Mission ||
+				(pool == TextNotificationPool.Feedback && filters.HasFlag(TextNotificationPoolFilters.Feedback));
```
My tried brain thinks we can simplify this to: 
```
return (pool == TextNotificationPool.Feedback && filters.HasFlag(TextNotificationPoolFilters.Feedback)) || pool != TextNotificationPool.Feedback;
```

dragunoff replied:
https://github.com/OpenRA/OpenRA/commit/af04b4315c25527f5907b3a439a2a026871311cd#diff-1163f12a6d50e889e2f6219665a3415cdf0924803bc8a4a26e49ba0a0bf83563R69-R70 - I add another condition in #18929. And I found it simpler to list everything explicitly.

teinarss commented on OpenRA.Game/TextNotificationsManager.cs:
```diff
@@ -28,24 +29,44 @@ static TextNotificationsManager()
 		}
 
-		public static void AddChatLine(string name, Color nameColor, string text)
+		public static void AddChatLine(string prefix, string text, Color? prefixColor = null, Color? textColor = null, TextNotificationPool? pool = null)
```
Rename to `AddTextNotification`?

dragunoff replied:
Renamed this and a bunch of other things that referenced "chat lines".

dragunoff commented on OpenRA.Game/Settings.cs:
```diff
@@ -33,6 +33,13 @@ public enum MPGameFilters
+	public enum TextNotificationPoolFilters
+	{
+		None = 0,
+		Feedback = 1
```
Perhaps this pool should be named "Transients" after all :thinking: I'm considering taking a shot at implementing a transient messages widget (discussed in #16355). "Feedback" and "UI Feedback" feels quite limited and is not very clear.

dragunoff replied:
After mulling over this for some more and watching some CNC3 footage I decided to leave it at "Feedback". In a follow up PR I plan to attempt a transient messages widget that combines UI feedback and transcriptions, similar to this:
![image](https://user-images.githubusercontent.com/1355810/121682558-e89df980-cac4-11eb-91da-fd82a4fdc8b9.png)

dragunoff replied:
See #19458

dragunoff commented:
Update: Addressed some feedback by @teinarss and shuffled the methods in `TextNotificationsManager` some more. This has taken many iterations but I feel it's finally coming together :crossed_fingers:

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs:
```diff
@@ -231,38 +234,58 @@ public void CloseChat()
+				chatLineToDisplay = new TextNotification(
+					chatLine.Pool,
+					chatLine.Prefix,
+					string.Format("{0} ({1})", chatLine.Text, repetitions + 1),
```
Can now be `$"{chatLine.Text} ({repetitions + 1})"`

pchote commented on OpenRA.Game/TextNotification.cs:
```diff
@@ -0,0 +1,56 @@
+		public readonly Color PrefixColor;
+		public readonly string Prefix;
+		public readonly string Text;
+		public readonly Color TextColor;
```
Pedantic nit (feel free to ignore): we list (color, text, text, color) here compared to (text, text, color, color) in the ctor. Could we reorder these fields to either (text, color, text, color) or (text, text, color, color) for consistency?

pchote commented on OpenRA.Game/TextNotificationsManager.cs:
```diff
@@ -28,24 +28,46 @@ static TextNotificationsManager()
+
+		static void AddTextNotification(TextNotificationPool pool, string prefix, string text, Color? prefixColor = null, Color? textColor = null)
+		{
+			var thePrefixColor = prefixColor ?? chatMessageColor;
```
Style nit: we don't use "the" prefixes anywhere else in the code so these stick out as looking a bit odd.
These two lines are quite short and only used once, so how about we just inline them directly below? That would also save us from evaluating them if the pool is not enabled.

pchote reviewed (commented):
LGTM, just a couple of very minor style nits then good to merge. I'm looking forward to this and #19458.

dragunoff commented:
Update: addressed the nits from pchote

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/3ea046323dd9f25bd71803cad917f09557befb78
